### ISSUE TITLE
causal convolutions

### ISSUE THREAD

**Issue by lucidrains** [lucidrains/audiolm-pytorch#8]
![Screenshot from 2022-10-26 09-35-39](https://user-images.githubusercontent.com/108653/198083735-96e4fdf1-a4d9-47dc-badd-1c3db56e8d4b.png)

in the [soundstream paper](https://arxiv.org/pdf/2107.03312.pdf), they claim all the convolutions are causal, but does this apply to the upsampling convtranspose? i'm not sure if it is possible to maintain causality

**Comment by lucidrains** [comment]
@NPN omg, thank you! and thank you for taking the time to make the diagram!

**Comment by NPN** [comment]
No problem! Regarding https://github.com/lucidrains/audiolm-pytorch/commit/d9c6ba4053eba6ddaa0578f5eff499b7663f2169, though, I don't think the padding (`x = F.pad(x, (self.padding, 0))`) is needed, since it makes the first `stride * (kernel_size - 1)` outputs equal to just the bias. Taking this code:

```python
c = CausalConvTranspose1d(chan_in=1, chan_out=1, kernel_size=4, stride=2)
nn.init.ones_(c.conv.weight)
nn.init.zeros_(c.conv.bias)
x = torch.tensor([1.0, 2.0, 4.0, 8.0]).reshape(1, 1, -1)
c(x)
```
With padding, we get `[0, 0, 0, 0, 0, 0, 1, 1]`, where the first 6 elements are just the bias. Without padding, we get `[1, 1, 3, 3, 6, 6, 12, 12]`.

**Comment by lucidrains** [comment]
@NPN yes you are correct, i reverted to that code and forgot to remove it :bow:
